File: app-go/internal/user/infrastructure/web/two_factor_controller.go
package web

import (
	"net/http"

	"github.com/gin-gonic/gin"
	"github.com/google/uuid"
	"github.com/orkhan/ecommerce/internal/middleware"
	"github.com/orkhan/ecommerce/internal/shared/domain"
	"github.com/orkhan/ecommerce/internal/shared/infrastructure/api"
	"github.com/orkhan/ecommerce/internal/shared/infrastructure/auth"
	userDomain "github.com/orkhan/ecommerce/internal/user/domain"
)

// TwoFactorController — 2FA TOTP endpoint-ləri
//
// Laravel: app/Http/Controllers/TwoFactorController.php
// Spring: TwoFactorController.java
type TwoFactorController struct {
	repo       userDomain.Repository
	twoFactor  *auth.TwoFactorService
	jwtService *auth.JWTService
}

func NewTwoFactorController(repo userDomain.Repository, tf *auth.TwoFactorService, jwt *auth.JWTService) *TwoFactorController {
	return &TwoFactorController{repo: repo, twoFactor: tf, jwtService: jwt}
}

// POST /api/auth/2fa/enable
func (c *TwoFactorController) Enable(ctx *gin.Context) {
	uid := ctx.MustGet(middleware.UserIDKey).(uuid.UUID)
	user, err := c.repo.FindByID(userDomain.UserID(uid))
	if err != nil || user == nil {
		ctx.Error(domain.NewEntityNotFoundError("User", uid.String()))
		return
	}

	setup, err := c.twoFactor.GenerateSecret(user.Email().Value())
	if err != nil {
		ctx.Error(err)
		return
	}
	ctx.JSON(http.StatusOK, api.SuccessWithMessage(gin.H{
		"secret":      setup.Secret,
		"qr_code_url": setup.OTPAuthURL,
	}, "QR-i scan edin və code ilə təsdiqləyin"))
}

// POST /api/auth/2fa/confirm  body: {secret, code}
func (c *TwoFactorController) Confirm(ctx *gin.Context) {
	uid := ctx.MustGet(middleware.UserIDKey).(uuid.UUID)
	var body struct {
		Secret string `json:"secret" binding:"required"`
		Code   string `json:"code" binding:"required"`
	}
	if err := ctx.ShouldBindJSON(&body); err != nil {
		ctx.Error(err)
		return
	}
	if !c.twoFactor.Verify(body.Secret, body.Code) {
		ctx.Error(domain.NewDomainError("2FA kodu yanlışdır"))
		return
	}

	user, err := c.repo.FindByID(userDomain.UserID(uid))
	if err != nil || user == nil {
		ctx.Error(domain.NewEntityNotFoundError("User", uid.String()))
		return
	}
	codes := c.twoFactor.GenerateBackupCodes(8)
	if err := user.EnableTwoFactor(body.Secret, codes); err != nil {
		ctx.Error(err)
		return
	}
	if err := c.repo.Save(user); err != nil {
		ctx.Error(err)
		return
	}
	ctx.JSON(http.StatusOK, api.SuccessWithMessage(codes,
		"2FA aktiv edildi. Backup kodları saxlayın!"))
}

// POST /api/auth/2fa/disable
func (c *TwoFactorController) Disable(ctx *gin.Context) {
	uid := ctx.MustGet(middleware.UserIDKey).(uuid.UUID)
	user, err := c.repo.FindByID(userDomain.UserID(uid))
	if err != nil || user == nil {
		ctx.Error(domain.NewEntityNotFoundError("User", uid.String()))
		return
	}
	user.DisableTwoFactor()
	if err := c.repo.Save(user); err != nil {
		ctx.Error(err)
		return
	}
	ctx.JSON(http.StatusOK, api.SuccessWithMessage(nil, "2FA deaktiv edildi"))
}

// POST /api/auth/2fa/verify  body: {user_id, code}
func (c *TwoFactorController) Verify(ctx *gin.Context) {
	var body struct {
		UserID string `json:"user_id" binding:"required"`
		Code   string `json:"code" binding:"required"`
	}
	if err := ctx.ShouldBindJSON(&body); err != nil {
		ctx.Error(err)
		return
	}

	uid, err := uuid.Parse(body.UserID)
	if err != nil {
		ctx.Error(err)
		return
	}

	user, err := c.repo.FindByID(userDomain.UserID(uid))
	if err != nil || user == nil {
		ctx.Error(domain.NewEntityNotFoundError("User", body.UserID))
		return
	}
	if !c.twoFactor.Verify(user.TwoFactorSecret(), body.Code) {
		ctx.Error(domain.NewDomainError("2FA kodu yanlışdır"))
		return
	}

	token, err := c.jwtService.Issue(user.ID().UUID(), user.Email().Value())
	if err != nil {
		ctx.Error(err)
		return
	}
	ctx.JSON(http.StatusOK, api.Success(gin.H{"token": token}))
}

func (c *TwoFactorController) RegisterRoutes(public, authed *gin.RouterGroup) {
	public.POST("/auth/2fa/verify", c.Verify)
	authed.POST("/auth/2fa/enable", c.Enable)
	authed.POST("/auth/2fa/confirm", c.Confirm)
	authed.POST("/auth/2fa/disable", c.Disable)
}

File: app/spring/src/main/java/az/ecommerce/user/infrastructure/web/TwoFactorController.java
package az.ecommerce.user.infrastructure.web;

import az.ecommerce.shared.domain.exception.DomainException;
import az.ecommerce.shared.domain.exception.EntityNotFoundException;
import az.ecommerce.shared.infrastructure.api.ApiResponse;
import az.ecommerce.shared.infrastructure.auth.TwoFactorService;
import az.ecommerce.user.domain.User;
import az.ecommerce.user.domain.repository.UserRepository;
import az.ecommerce.user.domain.valueobject.UserId;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.Map;
import java.util.UUID;

/**
 * Laravel: app/Http/Controllers/TwoFactorController.php
 *
 * POST /api/auth/2fa/enable, /confirm, /disable, /verify
 */
@RestController
@RequestMapping("/api/auth/2fa")
public class TwoFactorController {

    private final UserRepository repository;
    private final TwoFactorService twoFactorService;
    private final JwtService jwtService;

    public TwoFactorController(UserRepository repository, TwoFactorService twoFactorService,
                               JwtService jwtService) {
        this.repository = repository;
        this.twoFactorService = twoFactorService;
        this.jwtService = jwtService;
    }

    @PostMapping("/enable")
    public ApiResponse<Map<String, Object>> enable(@AuthenticationPrincipal UUID userId) {
        User user = repository.findById(new UserId(userId))
                .orElseThrow(() -> new EntityNotFoundException("User", userId.toString()));
        var setup = twoFactorService.generateSecret(user.email().value());
        return ApiResponse.success(Map.of(
                "secret", setup.secret(),
                "qr_code_url", setup.otpAuthUrl()
        ), "Bu QR-i scan edin və code ilə təsdiq edin");
    }

    @PostMapping("/confirm")
    public ApiResponse<List<String>> confirm(@AuthenticationPrincipal UUID userId,
                                              @RequestBody Map<String, Object> body) {
        String secret = (String) body.get("secret");
        Object rawCode = body.get("code");
        if (rawCode == null) throw new DomainException("'code' sahəsi tələb olunur");
        int code = Integer.parseInt(rawCode.toString());

        if (!twoFactorService.verifyCode(secret, code)) {
            throw new DomainException("2FA kodu yanlışdır");
        }

        User user = repository.findById(new UserId(userId))
                .orElseThrow(() -> new EntityNotFoundException("User", userId.toString()));
        List<String> backupCodes = twoFactorService.generateBackupCodes(8);
        user.enableTwoFactor(secret, backupCodes);
        repository.save(user);

        return ApiResponse.success(backupCodes, "2FA aktiv edildi. Backup kodları saxlayın!");
    }

    @PostMapping("/disable")
    public ApiResponse<Void> disable(@AuthenticationPrincipal UUID userId) {
        User user = repository.findById(new UserId(userId))
                .orElseThrow(() -> new EntityNotFoundException("User", userId.toString()));
        user.disableTwoFactor();
        repository.save(user);
        return ApiResponse.success(null, "2FA deaktiv edildi");
    }

    @PostMapping("/verify")
    public ApiResponse<Map<String, String>> verify(@RequestBody Map<String, Object> body) {
        UUID userId = UUID.fromString((String) body.get("user_id"));
        Object rawCode = body.get("code");
        if (rawCode == null) throw new DomainException("'code' sahəsi tələb olunur");
        int code = Integer.parseInt(rawCode.toString());

        User user = repository.findById(new UserId(userId))
                .orElseThrow(() -> new EntityNotFoundException("User", userId.toString()));

        if (!twoFactorService.verifyCode(user.twoFactorSecret(), code)) {
            throw new DomainException("2FA kodu yanlışdır");
        }

        String token = jwtService.issue(user.id().value(), user.email().value());
        return ApiResponse.success(Map.of("token", token));
    }
}
